server/src/transport/ws.rs

```diff
@@ -268,68 +295,128 @@ async fn send_task(
-			Either::Left((Either::Left((Some(Err(e)), _)), s)) => break Receive::Err(e, s),
-			// Max inactivity timeout fired, check if the connection has been idle too long.
-			Either::Left((Either::Right((_instant, rcv)), s)) => {
-				if let Some(p) = ping_config {
-					if last_active.elapsed() > p.inactive_limit {
-						*missed_pings += 1;
-
-						if *missed_pings >= p.max_failures {
-							tracing::debug!(
-								target: LOG_TARGET,
-								"WS ping/pong inactivity limit `{}` exceeded; closing connection",
-								p.max_failures,
-							);
-							break Receive::ConnectionClosed;
-						}
+			Either::Left((Either::Left((Some(Err(e)), s)), _)) => break Receive::Err(e, s),
+
+			// Server has been stopped or closed by inactive peer.
+			Either::Left((Either::Right((_, _)), _)) => break Receive::Stopped,
+			// Ping task has been stopped.
+			Either::Right((_, _)) => break Receive::KeepAliveExpired,
+		}
+	}
+}
+
+#[derive(Debug, Copy, Clone)]
+enum KeepAlive {
+	Ping(Instant),
+	Data(Instant),
+	Pong(Instant),
+}
+
+async fn ping_pong_task(
+	mut rx: mpsc::Receiver<KeepAlive>,
+	max_inactive_limit: Duration,
+	max_inactive: usize,
+	remote_addr: Option<IpAddr>,
+) {
+	let mut polling_interval = IntervalStream::new(interval(max_inactive_limit));
+	let mut pending_pings: VecDeque<Instant> = VecDeque::new();
+	let mut missed_pings = 0;
+
+	loop {
+		tokio::select! {
+			// If the ping is never answered, we use this timer as a fallback.
+			_ = polling_interval.next() => {
+				let mut remove = false;
+
+				if let Some(ping_start) = pending_pings.front() {
+					let elapsed = ping_start.elapsed();
+
+					if elapsed >= max_inactive_limit {
+						missed_pings += 1;
+						remove = true;
+						tracing::debug!(target: LOG_TARGET, "Ping/pong keep alive expired for peer={:?}, elapsed={}ms/max={}ms", remote_addr, elapsed.as_millis(), max_inactive_limit.as_millis());
+					}
+
+					if missed_pings >= max_inactive {
+						tracing::debug!(target: LOG_TARGET, "Missed {missed_pings} ping/pongs for peer={:?}; closing connection", remote_addr);
+						break;
 					}
 				}
 
-				stopped = s;
-				futs = futures_util::future::select(rcv, inactivity_check.next());
+				if remove {
+					pending_pings.pop_front();
+				}
+			}
+			msg = rx.recv() => {
+				match msg {
+					Some(KeepAlive::Ping(start)) => {
+						pending_pings.push_back(start);
+					}
+					Some(KeepAlive::Pong(end)) | Some(KeepAlive::Data(end)) => {
+						// Both pong and data are considered as a response to the ping.
+						// So we might get more responses than pings that's why it's possible
+						// that the pending_pings may be empty.
+						if let Some(start) = pending_pings.pop_front() {
+							// Calculate the round-trip time (RTT) of the ping/pong.
+							// We adjust for the time it took to send to this task.
+							let elapsed = start.elapsed().saturating_sub(end.elapsed());
+
+							if elapsed >= max_inactive_limit {
+								missed_pings += 1;
```

Comment:
What decrements this again? If somebody misses a few pings but then is ok and replies to the next 100 pings after that, I'd expect this counter to be reset to 0 again, else it might gradually accumulate until it hits the limit.

I'd prob reset it each time a successful ping comes in, eg an `else` on this?